src/Avalonia.Base/ValueStore.cs

```diff
@@ -178,5 +195,114 @@ public DeferredSetter<object> Setter
                     (_deferredSetter = new DeferredSetter<object>());
             }
         }
+
+        private bool TryGetValue(AvaloniaProperty property, out object value)
+        {
+            (int index, bool found) = TryFindEntry(property.Id);
+            if (!found)
+            {
+                value = null;
+                return false;
+            }
+
+            value = _entries[index].Value;
+            return true;
+        }
+
+        private void AddValueInternal(AvaloniaProperty property, object value)
+        {
+            Entry[] entries = new Entry[_entries.Length + 1];
+
+            for (int i = 0; i < _entries.Length; ++i)
+            {
+                if (_entries[i].PropertyId > property.Id)
+                {
+                    if (i > 0)
+                    {
+                        Array.Copy(_entries, 0, entries, 0, i);
+                    }
+
+                    entries[i] = new Entry { PropertyId = property.Id, Value = value };
+                    Array.Copy(_entries, i, entries, i + 1, _entries.Length - i);
```

Comment:
Why use an array and manually copy here instead of using a `List<>`?

Reply:
1. List<> has internal fields that consume 12 bytes (32-aligned) or 24 bytes (64-aligned) of memory (in addition to the internal raw array).
2. Accessing a list item may potentially be a double indirection instead of a single indirection.
3. It also contains an additional if clause within the indexer accessor.

However, I agree that using a raw array is quite verbose.